### PR TITLE
lib: json: Fix crashes on xtensa/nios2

### LINKED ISSUE #1347: json test fails on both xtensa and nios2
Test is enabled to build only and not run in Qemu, when we enable this to run in Qemu (CI), both architectures fail

```
core:json(relnotes): less ../../../sanity-out/qemu_xtensa/json/test/qemu.log
***** BOOTING ZEPHYR OS v1.9.0-rc2 - BUILD: Sep  4 2017 13:13:29 *****
Running test suite lib_json_test
===================================================================
starting test - test_json_encoding

    Assertion failed at /home/nashif/Work/github/zephyr/tests/lib/json/src/main.c:129: test_json_encoding: Encoded contents consistent (ret not equal to 0)
FAIL - test_json_encoding.
===================================================================
starting test - test_json_decoding
PASS - test_json_decoding.
===================================================================
starting test - test_json_obj_arr_encoding

    Assertion failed at /home/nashif/Work/github/zephyr/tests/lib/json/src/main.c:231: test_json_obj_arr_encoding: Encoded array of objects is consistent (!strcmp(buffer, encoded) is false)
FAIL - test_json_obj_arr_encoding.

```

### PR DESCRIPTION
**Pull request by lpereira**
This fixes two issues: alignment not being considered when calculating struct size, and an off-by-one error when serializing an object.

Tested on Linux with both address and behavior sanitizers, and on Zephyr running on qemu_nios2 and qemu_xtensa.

Should fix #1347.

### CODE REVIEW

**Review by nashif — CHANGES_REQUESTED** [review]
remove build_only: true in testcase.yaml, this would verify the fix.

**Comment by lpereira** [comment]
@mbolivar Do you have any idea how the alignment issue could be fixed? I'm not really happy with this way here; it's quite wasteful.

**Review by tyrantlizard77 — DISMISSED** [review]
I'm curious about the warnings due to use of anonymous types, since sanitycheck builds with -Werror in what situations do the warnings appear?

**Comment by andyross** [comment]
Seems sane.  One question.

**Comment by lpereira** [comment]
@andyross What's the question?

@andrewboie I built this with `-Wextra` while debugging on Linux, so that's where the warning came from.

**Review by andyross — DISMISSED** [review]
(Sorry, left my inline comment in a "review" and then posted a new "comment" leaving this stranded)

File: lib/json/json.c [hunk 1]
```diff
@@ -489,8 +491,14 @@ static ptrdiff_t get_elem_size(const struct json_obj_descr *descr)
 		ptrdiff_t total = 0;
 		size_t i;
 
-		for (i = 0; i < descr->sub_descr_len; i++) {
-			total += get_elem_size(&descr->object.sub_descr[i]);
+		for (i = 0; i < descr->object.sub_descr_len; i++) {
+			ptrdiff_t s = get_elem_size(&descr->object.sub_descr[i]);
+
+			if (s > (ptrdiff_t)descr->alignment) {
+				total += s;
```
[the remark this answers is not in the document]
Comment (lpereira) [reply lib/json/json.c]: Indeed. Pushed a new version that does `ROUND_UP(s, descr->alignment)` unconditionally.

**Review by andyross — APPROVED** [review]

**Pull request merged**